conbench/entities/benchmark_result.py

```diff
@@ -507,7 +459,115 @@ def ui_hardware_short(self) -> str:
         return f"{hw.id[:4]}: " + hw.name
 
 
-def do_iteration_samples_look_like_error(data: list[Optional[Decimal]]) -> bool:
+def validate_and_aggregate_samples(stats_usergiven: Any):
+    """
+    Raises BenchmarkResultValidationError upon logical inconsistencies.
+
+    Only run this for the 'success' case, i.e. when the input is a list longer
+    than zero, and each item is a number (not math.nan).
+
+    This returns a dictionary with key/value pairs meant for DB insertion,
+    top-level for BenchmarkResult.
+    """
+
+    # Encode invariants.
+    samples = stats_usergiven["data"]
+    assert len(samples) > 0
+    for sa in samples:
+        assert sa is not None
+        assert sa != math.nan
+
+    # Proceed with float type.
+    samples = [float(s) for s in samples]
+
+    # First copy the entire stats data structure (this includes times, data,
+    # mean, min, ...). Later: selectively overwrite/augment.
+    result_data_for_db = stats_usergiven.copy()
+
+    if len(samples) > 1:
+        # There is a special case where for one sample the iteration count
+        # may be larger than one, see below.
+        if stats_usergiven["iterations"] != len(samples):
+            raise BenchmarkResultValidationError(
+                f'iterations count ({ stats_usergiven["iterations"] }) does '
+                f"not match sample count ({len(samples)})"
+            )
+
+    if len(samples) >= 3:
+        # See https://github.com/conbench/conbench/issues/802 and
+        # https://github.com/conbench/conbench/issues/1118
+        q1, q3 = np.percentile(samples, [25, 75])
+
+        aggregates = {
+            "q1": q1,
+            "q3": q3,
+            "mean": np.mean(samples),
+            "median": np.median(samples),
+            "min": np.min(samples),
+            "max": np.max(samples),
+            # https://numpy.org/doc/stable/reference/generated/numpy.std.html
+            # This has N in the divisor (ddof=0)
+            "stdev": np.std(samples),
```

Comment:
Note: that's the same as before (see above in red part of diff: `stdev = np.std(dat)`).

But, most of the example data used in the test suite was built with ddof=1. That's where I believe most of the numerical differences come from that we see in the diff in the test suite.